src/media/media.ts

```diff
@@ -0,0 +1,1015 @@
+   *  in a "hardware mute" state. To see if a microphone _may_ be in a "hardware mute" 
+   *  state, you can listen for microphone volume events on 
+   *  `sdk.media.on('audioTrackVolume', evt)` and add logic to respond to no volume
+   *  coming through the microhpone. 
+   * 
+   * If `preserveMedia` is `true`, the media stream attained through the 
+   *  `startMedia()` will be returned to the caller. If not, the media will
+   *  be destroyed. 
+   * 
+   * `options` can be any valid deviceId or other media options defined in
+   *  `interface IMediaRequestOptions`. These options will be passed to 
+   *  the `startMedia()` call (which is used to gain permissions)
+   * 
+   * Note: the default option for the media type will be `true` (which is SDK default
+   *  device). If a value of `false` or `undefined` is passed in, it will 
+   *  always use `true`. Any options for the other media type will be ignored. 
+   * Example:
+   * ``` ts
+   * await requestMediaPermissions(
+   *   'audio',
+   *   false,
+   *   {
+   *     audio: false,
+   *     video: 'some-video-device-id',
+   *     videoFrameRate: 30
+   *   }  
+   * );
+   * // since type === 'audio', the options will be converted to: 
+   * {
+   *   // a valid option must be set (`false|undefined` are invalid)
+   *   audio: true,
+   *   // video will be ignored since permissions are requested one at a time
+   *   video: false
+   * }
+   * ```
+   * @param mediaType media type to request permissions for (`'audio' | 'video'`)
+   * @param preserveMedia flag to return media after permissions pass
+   * @param options optional, advanced options to request media with. 
+   */
+  async requestMediaPermissions (
+    mediaType: 'audio' | 'video',
+    preserveMedia = false,
+    options?: IMediaRequestOptions
+  ): Promise<MediaStream | void> {
+    options = options || {};
+    const oppositeMediaType = mediaType === 'audio' ? 'video' : 'audio';
+    const permissionsReqKey: keyof SdkMediaState = mediaType === 'audio'
+      ? 'micPermissionsRequested'
+      : 'cameraPermissionsRequested';
+    const permissionsKey: keyof SdkMediaState = mediaType === 'audio'
+      ? 'hasMicPermissions'
+      : 'hasCameraPermissions';
+
+    /* first load devices */
+    await this.enumerateDevices();
+
+    this.setPermissions({ [permissionsReqKey]: true });
+    /* make sure the options are valid */
+    if (options[mediaType] === undefined || options[mediaType] === false) {
+      options[mediaType] = true;
+    }
+    options[oppositeMediaType] = false;
+
+    /* delete the session off this before logging */
+    const optionsToLog = {
+      ...options,
+      sessionId: options?.session?.id,
+      conversationId: options?.session?.conversationId
+    };
+    delete optionsToLog.session;
+
+
+    this.sdk.logger.info('requesting media to gain permissions', { options: optionsToLog });
+
+    const stream = await this.startMedia(options);
```

Comment:
Does this create an infinite loop? Seems like it would.

Reply:
It won't because `startMedia()` calls here if `micPermissionsRequested | cameraPermissionsRequested` is `false`. This function will set those to `true` before calling `startMedia()`.

Reply:
Oh I see. 👍